### ISSUE TITLE
Allow type inference for class members

### ISSUE THREAD

**Issue by gafter** [dotnet/csharplang#2227]
@edward-a commented on [Tue Feb 12 2019](https://github.com/dotnet/roslyn/issues/33331)

I understand that using type inference for class members is a topic already raised in the past but here I want to show how the absence of it is highly detrimental to performance in some cases. In our code we have big auto-generated chunks with lots of lambdas passed as arguments. After removing all bottlenecks related to heap allocation, we are still stuck with delegates for lambdas heap-allocated upon instantiation of an object which significantly slows down the instantiation. We’d like to use class methods instead of lambdas called through static class methods to avoid the delegate allocations but in that case we would lose type inference which is no go for us. Please check the code below that presents the issue and the proposed solution.
``` c#
        class Class1
        {
            bool b;

            static double F(Class1 c) { return c.f(); }
            static Func<Class1, double> _F = F;
            double f() { return this.b ? 10.0 : 20.0; }

            public Class1()
            {
                Class2.Create(this, () => this.b ? 10.0 : 20.0); // Type inference but heap allocation (assume hundreds of this)

                Class2.Create(this, _F); // No heap allocation but can't use type inference for class members

                // Proposal:
                // static var F(Class1 c) { return c.f(); }
                // static var _F = F;
                // var f() { return this.b ? 10.0 : 20.0; }
            }
        }

        class Class2
        {
            public static Class3<T> Create<T>(Class1 c, Func<T> f)
            {
                return new Class3<T>(c, f);
            }

            public static Class3<T> Create<T>(Class1 c, Func<Class1, T> f)
            {
                return new Class3<T>(c, f);
            }

            public class Class3<T>
            {
                public Class3(Class1 c, Func<T> f) { }
                public Class3(Class1 c, Func<Class1, T> f) { }
            }
        }
```

**Comment by gafter** [comment]
How is the compiler supposed to know what delegate type to use for `static var _F = F;`?

**Comment by edward-a** [comment]
> 
> 
> How is the compiler supposed to know what delegate type to use for `static var _F = F;`?

I guess the same way it deduces the delegate type for a lambda? E.g. passing lambda as an argument in Class2.Create(this, () => this.b ? 10.0 : 20.0)

**Comment by gafter** [comment]
The compiler deduces the the delegate type for a lambda from the type it is converted to (e.g. a method parameter).  But in the declaration of `_F` it isn't converted to anything so that won't work.

**Comment by edward-a** [comment]
> The compiler deduces the the delegate type for a lambda from the type it is converted to (e.g. a method parameter). But in the declaration of `_F` it isn't converted to anything so that won't work.

The parameter of method is generic and also deduced by the compiler.

**Comment by gafter** [comment]
> The parameter of method is generic and also deduced by the compiler.

I have no idea what this refers to.

**Comment by edward-a** [comment]
> I have no idea what this refers to.
I updated the sample code to elaborate on what i mean. Notice that the version of the locally allocated delegate wrapper works just fine with type inference.  It would be great to have it also work for class members.

`    class Class1
    {
        bool b;

        static FuncDelegate<double> F = FuncDelegate.Create((Class1 c) => { return c.f(); });
        double f() { return this.b ? 10.0 : 20.0; }

        public Class1()
        {
            Class2.Create(this, () => this.b ? 10.0 : 20.0); // Type inference but heap allocation ode delegate (assume hundreds of these)

            Class2.Create(this, F.D); // No heap allocation but can't use type inference for class members

            // Proposal:
            // static var F = FuncDelegate.Create((Class1 c) => { return c.f(); });
            // var f() { return this.b ? 10.0 : 20.0; }

            // Compiler already can do that for local variables:
            var f = FuncDelegate.Create((Class1 c) => { return c.f(); });
            Class2.Create(this, f.D); // No heap allocation AND can use type inference!
        }
    }

    class Class2
    {
        public static Class3<T> Create<T>(Class1 c, Func<T> f)
        {
            return new Class3<T>(c, f);
        }

        public static Class3<T> Create<T>(Class1 c, Func<Class1, T> f)
        {
            return new Class3<T>(c, f);
        }

        public class Class3<T>
        {
            public Class3(Class1 c, Func<T> f) { }
            public Class3(Class1 c, Func<Class1, T> f) { }
        }
    }

    class FuncDelegate
    {
        public static FuncDelegate<T> Create<T>(Func<Class1, T> D)
        {
            return new FuncDelegate<T>(D);
        }
    }

    class FuncDelegate<T>
    {
        public Func<Class1, T> D;

        public FuncDelegate(Func<Class1, T> D)
        {
            this.D = D;
        }
    }
`

**Comment by HaloFour** [comment]
I don't understand the point of your example.  Why are you wrapping delegates with another class?

Inference of members has been discussed to death.  It's not worth the cost to implement:

https://github.com/dotnet/csharplang/issues/100#issuecomment-303187152

**Comment by edward-a** [comment]
> I don't understand the point of your example. Why are you wrapping delegates with another class?
I did it to use type inference that is already available in the compiler for local variables, but this time for a class member.

> Inference of members has been discussed to death. It's not worth the cost to implement:
I my case, type inference for class members allows to through out delegate allocation for lambdas and use methods instead. This is critical as it allows optimization of code parts that have many of lambdas passed as arguments and where the type inference is still needed. So, you believe that re-using lambda's return type inference for class methods to give that kind of flexibility in optimization is not worth it or too difficult to implement?

**Comment by HaloFour** [comment]
> So, you believe that re-using lambda's return type inference for class methods to give that kind of flexibility in optimization is not worth it or too difficult to implement?

I think that you can't enable inference for a single scenario.  The issues with member inference are due to the fact that, unlike locals, fields can forward reference.  With lambdas it's even worse because a lambda has no type, the compiler must use the target type of the assignment to figure out the delegate type to use.

There's nothing about the language today that forces you to incur additional allocations, you just have to be more explicit with your generic type arguments.

**Comment by edward-a** [comment]
> There's nothing about the language today that forces you to incur additional allocations, you just have to be more explicit with your generic type arguments.

Well, i could be wrong, but doesn't a call like 'new a(() => this.b + this.c);' result in delegate allocation on heap?

> I think that you can't enable inference for a single scenario. The issues with member inference are due to the fact that, unlike locals, fields can forward reference. With lambdas it's even worse because a lambda has no type, the compiler must use the target type of the assignment to figure out the delegate type to use.

The minimum needed to avoid delegate allocation in our case would be having 'var' as a return type of class method that i guess would be similar in type inference of lambda's result? Isn't it possible to only enable type inference for method's return type?

**Comment by HaloFour** [comment]
> Well, i could be wrong, but doesn't a call like 'new a(() => this.b + this.c);' result in delegate allocation on heap?

Delegate instances involve allocations.  So do closures.  What does that have to do with type inference?

> The minimum needed to avoid delegate allocation in our case would be having 'var' as return type

This makes no sense.  Type inference doesn't change what the code does.  The only time `var` is ever necessary in C# is for unspeakable types, like anonymous classes.

> Isn't it possible to only enable type inference for method's return type?

No.  That method may be used by another method that infers the return type, and so on, and so on, and so on.

**Comment by edward-a** [comment]
> Delegate instances involve allocations. So do closures. What does that have to do with type inference?

This particular example is not to demonstrate inference, but to get confirmation that delegate allocation occurs.

> This makes no sense. Type inference doesn't change what the code does. The only time `var` is ever necessary in C# is for unspeakable types, like anonymous classes.

Type inference of method is needed to avoid delegate allocation in the constructor (see example in the beginning of this thread) and instead allocate it statically and use it as a proxy to call class method. When you have performance critical code with dozens of lambdas it pretty much makes sense to have means to reduce heap allocation.

> No. That method may be used by another method that infers the return type, and so on, and so on, and so on.

That makes sense and i am not ready to present a counter-argument.

**Comment by HaloFour** [comment]
> Type inference of method is needed to avoid delegate allocation in the constructor

I don't understand this assertion at all, and I can't really make heads or tails of what your example is trying to demonstrate.

That said, I think you misinterpret what the C# compiler is doing for your first example.  It doesn't need to create an allocation for the lambda because you are not enclosing any additional state.  You still incur the allocation for the delegate itself, but that's the cost of using delegates.

**Comment by edward-a** [comment]
> I don't understand this assertion at all, and I can't really make heads or tails of what your example is trying to demonstrate.

Let me summarize it for you:
1. I need to avoid delegate allocations for lambdas on every instance of an object that uses lambdas a lot.
2. For that i can use class method instead of lambda called from a static method that has its delegate statically allocated and accessible as a class member.
3. I need to maintain type inference in the above as with lambdas.

Makes sense?

**Comment by HaloFour** [comment]
@edward-a 

Not really, no.  Type inference of members doesn't change how delegates or allocations work.

**Comment by edward-a** [comment]
> Not really, no. Type inference of members doesn't change how delegates or allocations work.

Who said it does? Type interference is needed to create generic objects as shown in the sample code in the beginning of this thread:

````csharp 
Class2.Create(this, () => this.b ? 10.0 : 20.0); // Type inference but heap allocation (assume hundreds of this)

Class2.Create(this, _F); // No heap allocation but can't use type inference for class members
`

**Comment by yaakov-h** [comment]
> We’d like to use class methods instead of lambdas called through static class methods to avoid the delegate allocations but in that case we would lose type inference which is no go for us.

\- OP

Why is explicitly declaring function/delegate types a hard "no go"?

`var` for method returns etc. has been discussed previously and closed off as "no we will not do this". In your proposed example, from which I've stripped the leading comments:

```csharp
// Proposal:
static var F(Class1 c) { return c.f(); }
static var _F = F;
var f() { return this.b ? 10.0 : 20.0; }
```

In order for the compiler to select a type for `_F`, it has to look at the return type for `F`. To do that, it has to inspect `c.f()`, matching any overloads, then has to infer the return type of that as well, which in this simple case can easily be inferred as `double`.

This is problematic, impossible to solve in some scenarios, and virtually on the list of things that will never be done.

https://github.com/dotnet/csharplang/issues/1484#issuecomment-394531995

**Comment by HaloFour** [comment]
And ultimately

```cs
static var _F = F;
```

Is impossible because the compiler can never know what `_F` is supposed to be.  Lambdas have to resolve to a delegate type and the compiler deduces that based on assignment.

**Comment by edward-a** [comment]
> Why is explicitly declaring function/delegate types a hard "no go"?

Because the code is auto generated from DSL that contains expressions with inferred return types. There is no way around this.

> `var` for method returns etc. has been discussed previously and closed off as "no we will not do this". 

Ok.

> This is problematic, impossible to solve in some scenarios, and virtually on the list of things that will never be done.

Understood.

**Comment by edward-a** [comment]
> ```cs
> static var _F = F;
> ```
> 
> Is impossible because the compiler can never know what `_F` is supposed to be. Lambdas have to resolve to a delegate type and the compiler deduces that based on assignment.

It has been already noted and changed here https://github.com/dotnet/csharplang/issues/2227#issuecomment-464404761

The static delegate is now wrapped and compiler has no problem inferring return type in case of a local variable.
```cs
// Proposal:
static var F = FuncDelegate.Create((Class1 c) => { return c.f(); });
var f() { return this.b ? 10.0 : 20.0; }
 ```

**Comment by CyrusNajmabadi** [comment]
> Because the code is auto generated from DSL that contains expressions with inferred return types. There is no way around this.

What does this mean?  What is this DSL?

**Comment by CyrusNajmabadi** [comment]
Why do you have `var f() { return this.b ? 10.0 : 20.0; }`?  why not just use `double f() { return this.b ? 10.0 : 20.0; }`?

**Comment by edward-a** [comment]
Gentlemen, it's really taxing to explain everything from the start to whoever joins this thread. Please read at least the beginning of it.

DSL - domain specific language
'var' is needed to replace lambda with method and maintain return type inference

**Comment by CyrusNajmabadi** [comment]
> It's really taxing to explain everything from the start to whoever joins this thread. Please read at least the beginning of it.

I've read all of this. 

> DSL - domain specific language

i know what a DSL is :)

I'm asking: why does your DSL insist on using 'var'?  Why not just actually put the appropriate types in the DSL?

You're basically askign C# to change for a totally arbitrary decision you made.  But i'm trying to understand that decision and why you can't cahnge, but the onus is entirely on us to change.

**Comment by edward-a** [comment]
> You're basically askign C# to change for a totally arbitrary decision you made. But i'm trying to understand that decision and why you can't cahnge, but the onus is entirely on us to change.

You have no idea why it's this way and already call it "totally arbitrary decision"? Jumping into conclusions much? The DSL is user-centered and specifying return type is excessive. Consider asking UWP team why they don't force to specify return type in x:Bind expressions in XAML.

**Comment by CyrusNajmabadi** [comment]
> You have no idea why it's this way and already call it "totally arbitrary decision"?

Yes.  

> The DSL is user-centered and specifying return type is excessive.

Then you need to figure out what that code needs to mean if the types are not supplied.  The user can supply whatever dsl is appropriate, and you can figure out a reasonable type to give for the 'vars'.  

You could take an approach like TypeScript takes in these cases to determine things.  That would unblock you without adding changes to the C# language that are felt to be undesirable.

**Comment by yaakov-h** [comment]
Could you not do something like:

1. Use DSL to generate code with lambdas
2. Parse the generated code to a syntax tree and semantic model
3. Rewrite the generated code to have static delegates to avoid lambda allocations

You'd basically just need a tool to look at your generated code, and hoist each lambda out into it's own static method + static field + instance method triplet.

**Comment by CyrusNajmabadi** [comment]
I agree with @yaakov-h .  The part you seem to be insistent on is that someone be able to just write the DSL and then have that be legal C# that you can just spit into a class.  There's no reason a DSL has to work this way.  You can control entirely how the DSL gets converted into C#.  So there's no need for the *end* C# to need to use 'var'.  You can actually just supply the right types based on whatever language rules you want your DSL to have.

This is why i was asking about DSL clarification questions.  Because it did seem exceptionally odd that you were both defining a dsl, but then requiring that DSL be something that would be inserted verbatim into C#.

**Comment by edward-a** [comment]
>     2. Parse the generated code to a syntax tree and semantic model
We can't build semantic model since we don't have type information when running our tool. It's a simple command line, not VS extension.

**Comment by edward-a** [comment]
> Then you need to figure out what that code needs to mean if the types are not supplied.

You are missing the point, cs compiler is pretty much capable to infer all the types in our case. It's not the issue. The issue is the performance hit we get with delegates dynamically allocated for lambdas with every instance of object.

**Comment by yaakov-h** [comment]
> We can't build semantic model since we don't have type information

> cs compiler is pretty much capable to infer all the types

I think you've tripped over yourself. Either there is sufficient type information, or there is not. Which one is it?

**Comment by edward-a** [comment]
> 
> 
> > We can't build semantic model since we don't have type information
> 
> > cs compiler is pretty much capable to infer all the types
> 
> I think you've tripped over yourself. Either there is sufficient type information, or there is not. Which one is it?

There is a sufficient type information in VS compiler to deduce the types. But not in the tool that converts DSL to CS.

**Comment by yaakov-h** [comment]
so add Microsoft.CodeAnalysis to your tool and have fun. 😉

**Comment by edward-a** [comment]
> so add Microsoft.CodeAnalysis to your tool and have fun. 😉

Consider i did that, how am i supposed to build semantic model if there is no type information available in the tool in relation to the project being built by VS? Those would be referenced assemblies, projects, and types declared in the project that contains auto-generated code. We currently avoid all of this by relying on VS' cs-compiler and it works great. But alas has its performance bottlenecks with delegates allocated with every instance. This is the problem i am eager to solve and using 'var' is just a proposed solution to it.

**Comment by CyrusNajmabadi** [comment]
> We can't build semantic model since we don't have type information when running our tool. It's a simple command line, not VS extension.

You don't need to be in VS to get type-information.  The Roslyn analysis APIs can be used entirely as a library that you could use in a command line app.

**Comment by CyrusNajmabadi** [comment]
I'm still lost on this: 

>> How is the compiler supposed to know what delegate type to use for `static var _F = F;`?

> I guess the same way it deduces the delegate type for a lambda? E.g. passing lambda as an argument in Class2.Create(this, () => this.b ? 10.0 : 20.0)

>> The compiler deduces the the delegate type for a lambda from the type it is converted to (e.g. a method parameter). But in the declaration of `_F` it isn't converted to anything so that won't work.

> The parameter of method is generic and also deduced by the compiler.

>> I have no idea what this refers to.

i don't know what it means either.  Today, the compiler figures out what a lambda means by trying to convert it to some *known* delegate type.  That's the *only* way it works today.  I don't see how you get around that.  There is no case in the language today where we support `var _F = SomeMethodGroup`.  We don't support it *inside* a method, so it's practically certain we can't support it *outside* a method.

**Comment by edward-a** [comment]
> 
> 
> > We can't build semantic model since we don't have type information when running our tool. It's a simple command line, not VS extension.
> 
> You don't need to be in VS to get type-information. The Roslyn analysis APIs can be used entirely as a library that you could use in a command line app.

You have to specify referenced assemblies when creating semantic model. Those are not only standard assemblies but also custom ones and those produced by projects in the solution.

**Comment by yaakov-h** [comment]
You can use Microsoft.CodeAnalysis.Workspaces.MSBuild to load in an entire solution, project references and all.

**Comment by CyrusNajmabadi** [comment]
> We currently avoid all of this

As mentioned above, your arbitrary decisions should not be used to drive the direction of hte language here.  You have *chosen* to try to write a tool that says "i don't want to understand types, and i want people to write arbitrary C# exprs, and i hope the C# language will figure it out for me".  But your initial decision is hamstringing you.  

Instead of writing your tool this way, you could just have hte language work the way it does today, but have your tool operate in a smarter fashion, using the Roslyn APIs to analyze the code and then do whatever it is you think is sensible for your domain.  As the purpose here is a *domain specific* language, it woudl make sense that you woudl have your own domain-specific semantics that you would need to map to C#.  

After all, if i write a DSL where it's legal to do `+ 1 1` (i.e. ReversePolish), then the onus isn't on the C# team to just make C# support that code if i paste it into a file.  Instead, the onus is on me to convert my DSL to the appropriate C# equivalent that has the semantics i want.

**Comment by CyrusNajmabadi** [comment]
> You have to specify referenced assemblies when creating semantic model. Those are not only standard assemblies but also custom ones and those produced by projects in the solution.

All of this can be done with public APIs provided by Roslyn.

**Comment by edward-a** [comment]
> You can use Microsoft.CodeAnalysis.Workspaces.MSBuild to load in an entire solution, project references and all.

This will drastically the build time. Imagine solution being loaded every time our tool runs on build and semantic model being built. We currently have almost no overhead translating DSL to CS.

**Comment by edward-a** [comment]
> As mentioned above, your arbitrary decisions should not be used to drive the direction of hte language here. You have _chosen_ to try to write a tool that says "i don't want to understand types, and i want people to write arbitrary C# exprs, and i hope the C# language will figure it out for me". But your initial decision is hamstringing you.

You do realize that we have no problem translating DSL to CS without type info and this is not the issue?

**Comment by CyrusNajmabadi** [comment]
> You do realize that we have no problem translating DSL to CS without type info and this is not the issue?

Then i've clearly missed the issue.  If you have no problem translating the dsl to cs without type info, and there's no issue with that... then why do you need any changes? I presumed there was an issue, which is why you wanted a language change.  

Can you succinctly specify what the issue is you're having as it seems to have gotten lost in the back and forth. 

Thanks!

**Comment by CyrusNajmabadi** [comment]
> This will drastically affect the build time. Imagine solution being loaded every time our tool runs when building and semantic model being built. 

Doesn't seem like much of issue. I've written lots of personal tools that use roslyn for this sort of thing, and the perf seems totally acceptable.  Indeed, this is how i often do investigations into language questions.  I'll write a little one-off console app.  Load many different solutions in it.  And then do whatever analysis i want.

> We currently have almost no overhead translating DSL to CS.

I can see why you'd want to preserve that.  But, ultimate, this was a choice you made in terms of how you would structure thins.  You both wanted to not every access semantics.  And you wanted to be able to allow people to have a DSL that cannot translate trivially into C#.  Your decisions here have led you to this point.

**Comment by edward-a** [comment]
> Then i've clearly missed the issue. If you have no problem translating the dsl to cs without type info, and there's no issue with that... then why do you need any changes?

The issue is the performance of the resulting cs code that can be improved as suggested.

**Comment by edward-a** [comment]
> And you wanted to be able to allow people to have a DSL that cannot translate trivially into C#. Your decisions here have led you to this point.

It does trivially translate to c# and that's the beauty of it. Consider it syntax sugar atop of c#.

**Comment by CyrusNajmabadi** [comment]
> It does trivially translate to c#

By 'c#' i meant: legal c#.

You seem to want it to be able to translate to something not legal today, but which you want to be legal.  So you're translating to some other language currently.

**Comment by CyrusNajmabadi** [comment]
> The issue is the performance of the resulting cs code that can be improved as suggested.

Could you try being a little less cryptic?  It's really unclear to me at this point what you actually want.  And core questions like https://github.com/dotnet/csharplang/issues/2227#issuecomment-465856974 remain unanswered.

I *still* at this point both don't have a strong idea of what you're asking for *and*, at the same time, have found any information i've been able to glean away very lacking in terms of how it would *actually* work.

can you simply and easily explain what it is you want *and* how you think it would work?  I've now read the entire thread *four* times (and have talked to someone else about this) and neither of us can actually condense out what the core request is.

**Comment by CyrusNajmabadi** [comment]
Note: this is being discussed over at https://gitter.im/dotnet/csharplang?at=5c6e0d68a7d733509d9db185.  Perhaps gitter would be a more conducive medium to figure out how to address yours scenario?

**Comment by edward-a** [comment]
> You seem to want it to be able to translate to something not legal today, but which you want to be legal. So you're translating to some other language currently.

I am currently translating to "legal" c#. Your legal/illegal taxonomy makes no sense to me, sorry.

**Comment by CyrusNajmabadi** [comment]
> I am currently translating to "legal" c#. Your legal/illegal taxonomy makes no sense to me, sorry.

You are asking for the language to support something it does not currently support.  That makes it illegal C#.  If it were legal, you wouldn't need a change to the language to support it.

Happy to discuss things more over on gitter.im.  it would likely help drive a positive conclusion faster than this back and forth where we're not making much progress and where many questions have been left unanswered.  Thanks!

**Comment by edward-a** [comment]
> Could you try being a little less cryptic? It's really unclear to me at this point what you actually want. 

Sure, let me try paraphrase the beginning of this thread. Consider we have the following code produced by our tool:

````cs
        class Class1
        {
            bool b;

            public Class1()
            {
                Class2.Create(this, () => this.b ? 10.0 : 20.0); // Type inference BUT heap allocation
            }
        }

        class Class2
        {
            public static Class3<T> Create<T>(Class1 c, Func<T> f)
            {
                return new Class3<T>(c, f);
            }

            public class Class3<T>
            {
                public Class3(Class1 c, Func<T> f) { }
            }
        }
````

It's nice and everything but we get delegate allocation in Class1 constructor. To get rid of it, we can do the following:

````cs
        class Class1
        {
            bool b;

            static double F(Class1 c) { return c.f(); }
            static Func<Class1, double> _F = F;
            double f() { return this.b ? 10.0 : 20.0; }

            public Class1()
            {
                Class2.Create(this, _F); // No heap allocation BUT can't use type inference for class members
            }
        }

        class Class2
        {
            public static Class3<T> Create<T>(Class1 c, Func<Class1, T> f)
            {
                return new Class3<T>(c, f);
            }

            public class Class3<T>
            {
                public Class3(Class1 c, Func<Class1, T> f) { }
            }
        }
````

Still nice and no delegate allocation but we lost our type inference.

**Comment by HaloFour** [comment]
> Still nice and no delegate allocation but we lost our type inference.

You can't have type inference there.  The C# compiler *requires* the delegate type in order to assign a method group or a lambda.  This is true with locals as well.

It looks like you have a solution that works just fine, it's just slightly more verbose than you'd prefer.  But for the many reasons already stated as to why member inference won't happen, compounded due to the fact that this involves inferring delegates, you're not likely to find language support specifically to benefit your DSL.  The benefactor becomes exactly one person.

**Comment by CyrusNajmabadi** [comment]
> `Class2.Create(this, () => this.b ? 10.0 : 20.0); // Type inference BUT heap allocation`

You have a heap allocation here because you're accessing `this`.  You don't do that in your other examples.  If you want to smply avoid the heap allocation here you can do that like so:

```c#
        class Class1
        {
            bool b;

            public Class1()
            {
                Class2.Create(this, c => c.b ? 10.0 : 20.0); // Type inference NO  heap allocation
            }
        }

        class Class2
        {
            public static Class3<T> Create<T>(Class1 c, Func<Class1, T> f)
            {
                return new Class3<T>(c, f);
            }

            public class Class3<T>
            {
                public Class3(Class1 c, Func<Class1, T> f) { }
            }
        }
```

**Comment by CyrusNajmabadi** [comment]
Note: the above allocates a delegate.  You could avoid that if you wanted pretty trivially.  Effectively, in the above code, you knew and spit out that the type was `Func<Class1, T> f`.  So, just 'lift' that into a member.  You already knew about this type in order for your original code to work.  So the only difference is that you make that type into a field, instead of just referencing it as a parameter.

**Comment by edward-a** [comment]
> You are asking for the language to support something it does not currently support. That makes it illegal C#. If it were legal, you wouldn't need a change to the language to support it.

Not sure what you are implying. Is it illegal to ask for a new feature? Or is there a set of features that are inherently illegal, will stay so indefinitely, and no one should mention them? If yes, i would prefer to have them listed somewhere, so i don't commit "crime" next time when posting. Your choice of words sounds dramatic but doesn't make much of a sense.

**Comment by HaloFour** [comment]
I will mention that Java has a feature that has been proposed for C# that may be of a benefit here, and that is method references.  It allows taking a reference to an instance method without requiring a target instance and treating the instance as the first parameter.  It doesn't eliminate the type inference (as Java has the same problem where functional interfaces are nominal), but it might simplify the other code generation you're doing.

```java
public class Class1 {
    public boolean b;
    public double f() {
        return this.b ? 10.0 : 20.0;
    }

    public static ToDoubleFunction<Class1> F = Class1::f;
}
```

**Comment by edward-a** [comment]
> You can't have type inference there. The C# compiler _requires_ the delegate type in order to assign a method group or a lambda. This is true with locals as well.

There is a way around this as shown in the second example in this thread:

````cs
    class Class1
    {
        bool b;

        double f() { return this.b ? 10.0 : 20.0; }

        public Class1()
        {
            var f = FuncDelegate.Create((Class1 c) => { return c.f(); });
            Class2.Create(this, f.D); // No heap allocation AND type inference.
        }
    }

    class Class2
    {
        public static Class3<T> Create<T>(Class1 c, Func<Class1, T> f)
        {
            return new Class3<T>(c, f);
        }

        public class Class3<T>
        {
            public Class3(Class1 c, Func<Class1, T> f) { }
        }
    }

    class FuncDelegate
    {
        public static FuncDelegate<T> Create<T>(Func<Class1, T> D)
        {
            return new FuncDelegate<T>(D);
        }
    }

    class FuncDelegate<T>
    {
        public Func<Class1, T> D;

        public FuncDelegate(Func<Class1, T> D)
        {
            this.D = D;
        }
    }
````

**Comment by CyrusNajmabadi** [comment]
> Your choice of words sounds dramatic but doesn't make much of a sense.

legal/illegal are simple and commonly used terms to describe programs that are allowed/disallowed by a language.  They don't have any moral connotation :)

**Comment by HaloFour** [comment]
@edward-a 

`FuncDelegate` is a heap allocation.

**Comment by edward-a** [comment]
> Note: the above allocates a delegate. You could avoid that if you wanted pretty trivially. Effectively, in the above code, you knew and spit out that the type was `Func<Class1, T> f`. So, just 'lift' that into a member. You already knew about this type in order for your original code to work. So the only difference is that you make that type into a field, instead of just referencing it as a parameter.

Not sure i am following. I would appreciate you showing it in code.

**Comment by edward-a** [comment]
> I will mention that Java has a feature that has been proposed for C# that may be of a benefit here, and that is method references. It allows taking a reference to an instance method without requiring a target instance and treating the instance as the first parameter. It doesn't eliminate the type inference (as Java has the same problem where functional interfaces are nominal), but it might simplify the other code generation you're doing.
> 
> ```java
> public class Class1 {
>     public boolean b;
>     public double f() {
>         return this.b ? 10.0 : 20.0;
>     }
> 
>     // the signature of ToDoubleFunction is:   double f(T arg1)
>     public static ToDoubleFunction<Class1> F = Class1::f;
> }
> ```

That would definitely be helpful, especially if you can also assign lambda as follows:

> ```java
> public class Class1 {
>     public boolean b;
> 
>     // the signature of ToDoubleFunction is:   double f(T arg1)
>     public static ToDoubleFunction<Class1> F = () => { return this.b ? 10.0 : 20.0; }
> }
> ```

**Comment by CyrusNajmabadi** [comment]
> Not sure i am following. I would appreciate you showing it in code.

Sure!

```C#
  class Class1
    {
        bool b;

        double f() { return this.b ? 10.0 : 20.0; }

        static Func<Class1, double> F = (Class1 c) => { return c.f(); };        

        public Class1()
        {
            Class2.Create(this, F); 
        }
    }
```

Basically, just as you knew to spit out a Func type that took `Class` in your example, just do the same here, but with the return type of whatever the 'return type of that lambda' is.

**Comment by edward-a** [comment]
> `FuncDelegate` is a heap allocation.

Yes, in case of local variable. The problem gets solved if the same can be done with a static variable.

**Comment by HaloFour** [comment]
> That would definitely be helpful, especially if you can also assign lambda as follows:

No, you can assign it as follows:

```java
public static ToDoubleFunction<Class1> F = (Class1 c) => { return c.b ? 10.0 : 20.0; }
```

Just like with C# you can't reference `this` within a static context.  But it's effectively the same thing as "this" is just the first parameter of an instance method anyway.

**Comment by CyrusNajmabadi** [comment]
@HaloFour Right.  that's what i was trying to express with: https://github.com/dotnet/csharplang/issues/2227#issuecomment-465866115

**Comment by edward-a** [comment]
> ```cs
>   class Class1
>     {
>         bool b;
> 
>         double f() { return this.b ? 10.0 : 20.0; }
> 
>         static Func<Class1, double> F = (Class1 c) => { return c.f(); };        
> 
>         public Class1()
>         {
>             Class2.Create(this, F); 
>         }
>     }
> ```
> 
> Basically, just as you knew to spit out a Func type that took `Class` in your example, just do the same here, but with the return type of whatever the 'return type of that lambda' is.
> 
> note: this would still be a single allocation. But presumably you're ok with that.

Thanks for the example, but i don't see a solution here the return type is still specified explicitly. Can you please elaborate?

**Comment by HaloFour** [comment]
> The problem gets solved if the same can be done with a static variable.

It can be done with a static variable, it just requires you to explicitly state the type.

> Thanks for the example, but i don't see a solution here the return type is still specified explicitly.

This is not going to change.  As you acknowledged in your first post, this has been brought up before.  Many times.  The team simply doesn't consider it remotely worth the effort, and there are numerous situations where the inference explodes due to forward referencing as has been mentioned here.  It's a very difficult problem with very little reward.  I don't see anything new or unique about your situation that would warrant reevaluating that position, especially since supporting your type-less DSL is not a goal of the C# language and the number of people who would benefit is vanishingly small.

**Comment by CyrusNajmabadi** [comment]
>Thanks for the example, but i don't see a solution here the return type is still specified explicitly. Can you please elaborate?

The solution is to specify the return type explicitly.  C# has no way around this.  And it's been a very well understood and intentional decision to not allow inference at the member level.  It's incredibly thorny and problematic (something i can speak to at length given the work we did in TypeScript).  The use case of "i want to have a dsl that is c#-like and would allow people to take pseudo-c# and paste it *literally* into c# and have it work without any augmentation" is a non-case.

C# cares little for DSLs.  It's fine with them existing, but the rule on them is: you need to translate your DSL into "legal" (i.e. 'correct') c#.  C# does not go out of its way to make arbitrary DSLs, with arbitrary rules, work unchanged. 

As i mentioned above (with the RPN case), if i came to c# and said "my DSL allows for `class Foo<> {}`, i want that to work in C#".  There would be no traction on that at all.  What i have there is my own DSL language and C# would only adopt it if it was valuable enough independently of my own needs and desires as a DSL author.

**Comment by edward-a** [comment]
> Just like with C# you can't reference `this` within a static context. But it's effectively the same thing as "this" is just the first parameter of an instance method anyway.

That feature with method references keeps me thinking. If lambdas don't capture any local variables, then they can be compiled into methods that can be referenced without delegate allocation?

**Comment by CyrusNajmabadi** [comment]
> If lambdas don't capture any local variables, then they can be compiled into methods that can be referenced without delegate allocation?

There have been discussions around this.  The main problem here is that this could be an observable change in behavior.  Consider, for example, someone who uses the delegate as a key into a dictionary.  Whether or not you get a new instance or not could change behavior.  

There are also security concerns here (thoughthey may be moot depending on the direction .net goes in) as delegates capture information about the stack that was used to create them, and those can be used to make security decisions.

I believe there's an open PR somewhere in Roslyn investigating this approach.  However, i think it's been years with very little traction.  So my guess is that it's not very palatable for the team.

**Comment by HaloFour** [comment]
> If lambdas don't capture any local variables, then they can be compiled into methods that can be referenced without delegate allocation?

You'd still need that initial allocation of the delegate, and you'd need to stick it somewhere that you can reference like a `static` field.

If you're asking if C# could selectively optimize the creation of certain kinds of delegates, yes it probably could, as long as there were no perceivable behavioral differences.  Those requests can be raised on the Roslyn repo since they wouldn't involve any changes to the C# grammar to support.

**Comment by edward-a** [comment]
> The solution is to specify the return type explicitly. C# has no way around this. 

The issues arising from having var as a return type of a method have already been mentioned in this thread. It's a good conversation to discover other solutions that might fit as in case of method references. For now, I can stay with the performance hit, it's an issue but not a show stopper.

**Comment by CyrusNajmabadi** [comment]
Ah, here it is: https://github.com/dotnet/roslyn/pull/6642

Looks like there's been some recent movement.  So maybe we'll see something come of this this year.

**Comment by edward-a** [comment]
> You'd still need that initial allocation of the delegate, and you'd need to stick it somewhere that you can reference like a `static` field.

Not sure i am following. Is my understanding correct that the bellow example doesn't ensue delegate allocation?

````cs
public class Class1 {
    public boolean b;
    public double f() {
        return this.b ? 10.0 : 20.0;
    }

    // the signature of ToDoubleFunction is:   double f(T arg1)
    public static ToDoubleFunction<Class1> F = Class1::f;
}
````

If yes, when why a lambda that can be translates to the very same method by the compiler would result in delegate allocation?

**Comment by CyrusNajmabadi** [comment]
> If yes, when why a lambda that can be translates to the very same method by the compiler would result in delegate allocation?

Because C# has no concept of any other way to store a reference to a method.  It's always in a delegate.

**Comment by edward-a** [comment]
> > If yes, when why a lambda that can be translates to the very same method by the compiler would result in delegate allocation?
> 
> Because C# has no concept of any other way to store a reference to a method. It's always in a delegate.

Then what are the benefits of introducing method references mentioned by @HaloFour

**Comment by CyrusNajmabadi** [comment]
@edward-a I think there's a lot of confusion in this discussion.  There seems to be a lot of conflation in your posts between what is actually possible in C# today versus what could potentially be possible with new language features.

**Comment by edward-a** [comment]
> @edward-a I think there's a lot of confusion in this discussion. There seems to be a lot of conflation in your posts between what is actually possible in C# today versus what could potentially be possible with new language features.

Indeed, it's quite a brainstorming here and i appreciate all the input. Thank you for the link about the delegate caching.  I am currently reading the thread struggling to understand the benefits of cached delegates.

**Comment by CyrusNajmabadi** [comment]
>  I am currently reading the thread struggling to understand the benefits of cached delegates.

The benefits of caching in this context are less allocations/garbage.

**Comment by jnm2** [comment]
It sounds like you came up with a design that requires a change to the C# language. To be realistic, let's imagine that the language design team really loved this idea and got to work on it right away. You'd be able to use it in C# 9 or 10, potentially five years from now. What would you do in the meantime, and what's the difference between doing it for only five years or doing it for the life of the project?
